core/network-libp2p/src/custom_proto/handler.rs

```diff
@@ -165,44 +158,199 @@ struct PerProtocolNormalState<TMessage, TSubstream> {
 	shutdown: SmallVec<[RegisteredProtocolSubstream<TMessage, TSubstream>; 4]>,
 }
 
-impl<TMessage, TSubstream> PerProtocol<TMessage, TSubstream>
+impl<TMessage, TSubstream> PerProtocolNormalState<TMessage, TSubstream>
 where TMessage: CustomMessage, TSubstream: AsyncRead + AsyncWrite {
-	/// Enables the protocol. Returns an optional event to emit.
-	/// Must be passed the endpoint of the connection.
-	#[must_use]
-	fn enable(&mut self, endpoint: Endpoint)
-		-> Option<ProtocolsHandlerEvent<RegisteredProtocol<TMessage>, ProtocolId, CustomProtosHandlerOut<TMessage>>> {
+	/// Polls for things that are new. Same API constraints as `Future::poll()`.
+	/// Optionally returns the event to produce.
+	/// You must pass the `protocol_id` as we need have to inject it in the returned event.
+	/// API note: Ideally we wouldn't need to be passed a `ProtocolId`, and we would return a
+	/// 	different enum that doesn't contain any `protocol_id`, and the caller would inject
+	/// 	the ID itself, but that's a ton of code for not much gain.
+	fn poll(&mut self) -> Option<CustomProtoHandlerOut<TMessage>> {
+		for n in (0..self.pending_response.len()).rev() {
+			let (request_id, mut substream) = self.pending_response.swap_remove(n);
+			match substream.poll() {
+				Ok(Async::Ready(Some(RegisteredProtocolEvent::Message(message)))) => {
+					if message.request_id() == CustomMessageId::Response(request_id) {
+						let event = CustomProtoHandlerOut::CustomMessage {
+							message
+						};
+						self.shutdown.push(substream);
+						return Some(event);
+					} else {
+						self.shutdown.push(substream);
+						let event = CustomProtoHandlerOut::ProtocolError {
+							is_severe: true,
+							error: format!("Request ID doesn't match substream: expected {:?}, \
+								got {:?}", request_id, message.request_id()).into(),
+						};
+						return Some(event);
+					}
+				},
+				Ok(Async::Ready(Some(RegisteredProtocolEvent::Clogged { .. }))) =>
+					unreachable!("Cannot receive Clogged message with new protocol version; QED"),
```

Comment:
As I'm reading the entire code for the first time it seem that it could be a bit more elaborate, might be good to give some hints (roughly) what parts of the code guarantees that.

Reply:
I was going to add a code comment, but it would be redundant with the message. Basically there's a backwards-compatible version of the protocol and a new version that we're going to enable in the future. The new version doesn't generate `Clogged` events.

Reply:
Removing the `Clogged` event is the next step in #1692.